proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+
+    data MyAppError = MissingConfigurationError | ...
+
+    readFile "my-app.conf" `catch` $ \ (ioe :: IOError) ->
+        if isDoesNotExistError ioe
+          then throwIO MissingConfigurationError
+          else throwIO ioe
+
+This pattern can be problematic in the presence of exception context: the
+exception thrown by the handler lacks any of the context attached to the
+original ``IOError``, including any backtraces.
+
+While in some select cases dropping context may be desireable (e.g. to avoid
+exposing implementation details unnecessarily to the user), in general this
+proposal seeks to make exception provenance information ubiquitous and
+reliable. Consequently, we propose to that ``catch`` and ``handle`` be modified
+to preserve exception context when an exception is thrown from a handler.
+
+
+Selecting backtrace mechanisms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the machinery described above, we can now address a common debugging
+scenario: locating the origin of an exception thrown by a third-party library.
+By far the most common means of throwing exceptions are ``throw``, ``throwIO``,
+``error``, and ``undefined``, none of which have any notion of backtrace collection.
+This raises the question of how the user should select which of the above
+mechanism(s) these functions should use to collect their backtrace.
+
+For this we propose a pragmatic, stateful approach to allow the user to enable
+individual mechanism(s) should be used for backtrace collection in ``throw``,
+``throwIO`` and similar functions: ::
+
+    module GHC.Exception.Backtrace
+        ( enableMechanism, BacktraceMechanism(..) ) where
+
+    -- | Which kind of backtrace to collect when an exception is thrown.
+    data BacktraceMechanism
+      = -- | collect a cost center stacktrace (only available when built with profiling)
+        CostCenterBacktraceMech
+      | -- | use execution stack unwinding with given limit
+        ExecutionStackBacktraceMech
+      | -- | collect backtraces from Info Table Provenance Entries
+        IPEBacktraceMech
+      | -- | use 'HasCallStack'
+        HasCallStackBacktraceMech
+      deriving (Eq, Show)
+
+    -- | Enable the given 'BacktraceMechanism' to be used when
+    -- 'Control.Exception.throwIO', et al. collect backtraces.
+    enableMechanism :: BacktraceMechanism -> IO ()
+    enableMechanism = ...
+        -- Internally this would be mutate program-global state
+
+    disableMechanism :: BacktraceMechanism -> IO ()
+    disableMechanism = ...
+
+A ``collectBacktrace`` primitive used by ``throwWithContext``
+simply dispatches to the currently-selected ``BacktraceMechanism``\ s: ::
+
+    module GHC.Exception.Backtrace where
+
+    -- | Collect an 'ExceptionContext' containing backtraces from all enabled
+    -- 'BacktraceMechanism's.
+    collectBacktraces :: HasCallStack => IO ExceptionContext
+    collectBacktraces = do
+        mechs <- readIORef currentBacktraceMechanisms
```

Comment:
Perhaps expose an API function from `GHC.Exception.Backtrace` to query the current list of backtrace mechanisms?